File: cortex/python/app/components/dispatcher.py
"""
Dispatcher Component

This module implements a dispatcher system that routes messages and events to appropriate
handlers. It serves as a central coordination point for component communication.
"""

import asyncio
import uuid
from datetime import datetime
from typing import Any, Callable, Dict, List, Optional, Set, Union

from pydantic import BaseModel, Field

from app.components.context_manager import ContextManager
from app.components.session_manager import SessionManager
from app.interfaces.domain_expert_interface import DomainExpertInterface
from app.utils.logger import get_contextual_logger

# Configure logger
logger = get_contextual_logger("components.dispatcher")

# Define handler type - async function that takes message and context
MessageHandler = Callable[[Dict[str, Any], Dict[str, Any]], Any]


class Dispatcher:
    """
    Dispatcher for routing messages and events

    This class manages message routing, handler registration, and coordination
    between different components of the system.
    """

    def __init__(
        self,
        context_manager: ContextManager,
        session_manager: SessionManager,
        domain_expert_interface: DomainExpertInterface,
    ):
        """
        Initialize the dispatcher

        Args:
            context_manager: The context manager instance
            session_manager: The session manager instance
            domain_expert_interface: The domain expert interface
        """
        self.context_manager = context_manager
        self.session_manager = session_manager
        self.domain_expert_interface = domain_expert_interface

        # Message handlers
        self.handlers: Dict[str, List[MessageHandler]] = {}

        # Active processors
        self.active_processors: Set[str] = set()

        logger.info("Dispatcher initialized")

    def register_handler(self, message_type: str, handler: MessageHandler) -> None:
        """
        Register a handler for a specific message type

        Args:
            message_type: Type of message to handle
            handler: Handler function
        """
        if message_type not in self.handlers:
            self.handlers[message_type] = []

        self.handlers[message_type].append(handler)
        logger.info(f"Registered handler for message type: {message_type}")

    async def dispatch(
        self,
        message_type: str,
        message: Dict[str, Any],
        context: Optional[Dict[str, Any]] = None,
        session_id: Optional[uuid.UUID] = None,
        workspace_id: Optional[uuid.UUID] = None,
    ) -> List[Dict[str, Any]]:
        """
        Dispatch a message to registered handlers

        Args:
            message_type: Type of message
            message: Message content
            context: Optional context information
            session_id: Optional session ID
            workspace_id: Optional workspace ID

        Returns:
            List of handler responses

        Raises:
            ValueError: If message type has no registered handlers
        """
        try:
            logger.info(f"Dispatching message: {message_type}")

            # Check if there are handlers for this message type
            if message_type not in self.handlers:
                logger.warning(
                    f"No handlers registered for message type: {message_type}"
                )
                return []

            # Initialize context if not provided
            if context is None:
                context = {}

            # Add processing metadata to context
            context.update(
                {
                    "message_type": message_type,
                    "dispatch_time": datetime.utcnow().isoformat(),
                    "dispatcher_id": id(self),
                }
            )

            # Add session info if provided
            if session_id:
                context["session_id"] = session_id

                # Get session info from session manager if available
                try:
                    session_info = await self.session_manager.get_session(session_id)
                    if session_info:
                        context["session_info"] = session_info.dict()
                except Exception as e:
                    logger.warning(f"Failed to get session info: {str(e)}")

            # Add workspace info if provided
            if workspace_id:
                context["workspace_id"] = workspace_id

                # Get context for this workspace if both session and workspace IDs are provided
                if session_id:
                    try:
                        workspace_context = await self.context_manager.get_context(
                            session_id=session_id,
                            workspace_id=workspace_id,
                        )
                        context["workspace_context"] = workspace_context.dict()
                    except Exception as e:
                        logger.warning(f"Failed to get workspace context: {str(e)}")

            # Process with all registered handlers
            results = []
            for handler in self.handlers[message_type]:
                try:
                    result = await handler(message, context)
                    if result:
                        results.append(result)
                except Exception as e:
                    logger.error(
                        f"Handler error for message type {message_type}: {str(e)}",
                        exc_info=True,
                    )

            logger.info(
                f"Message {message_type} processed by {len(self.handlers[message_type])} handlers"
            )
            return results

        except Exception as e:
            logger.error(
                f"Error dispatching message {message_type}: {str(e)}", exc_info=True
            )
            raise

    async def dispatch_with_expert(
        self,
        expert_type: str,
        task_type: str,
        content: Dict[str, Any],
        context: Optional[Dict[str, Any]] = None,
    ) -> str:
        """
        Dispatch a task to an expert via the domain expert interface

        Args:
            expert_type: Type of expert to use
            task_type: Type of task to perform
            content: Task content
            context: Optional context information

        Returns:
            Task ID for tracking

        Raises:
            ValueError: If expert delegation fails
        """
        try:
            logger.info(f"Dispatching task to expert: {expert_type}")

            # Create a task object
            from app.interfaces.domain_expert_interface import Task

            task = Task(
                type=task_type,
                content=content,
                context=context or {},
            )

            # Delegate task to expert
            task_id = await self.domain_expert_interface.delegate_task(
                expert_type=expert_type,
                task=task,
            )

            logger.info(f"Task delegated to expert {expert_type}, task ID: {task_id}")
            return task_id

        except Exception as e:
            logger.error(
                f"Error dispatching to expert {expert_type}: {str(e)}", exc_info=True
            )
            raise ValueError(f"Failed to dispatch to expert: {str(e)}")


# Global instance (will be initialized with required components later)
dispatcher = None


def initialize_dispatcher(
    context_manager: ContextManager,
    session_manager: SessionManager,
    domain_expert_interface: DomainExpertInterface,
) -> Dispatcher:
    """
    Initialize the global dispatcher instance

    Args:
        context_manager: The context manager to use
        session_manager: The session manager to use
        domain_expert_interface: The domain expert interface to use

    Returns:
        The initialized dispatcher
    """
    global dispatcher
    if dispatcher is None:
        dispatcher = Dispatcher(
            context_manager=context_manager,
            session_manager=session_manager,
            domain_expert_interface=domain_expert_interface,
        )
    return dispatcher


# Export public symbols
__all__ = [
    "Dispatcher",
    "MessageHandler",
    "dispatcher",
    "initialize_dispatcher",
]

File: cortex/node/src/components/dispatcher.ts
/**
 * Implementation of the Dispatcher component for Cortex Core
 */

import { v4 as uuidv4 } from 'uuid';
import { logger } from '../utils/logger';
import { ContextManager } from './context-manager';
import { DomainExpertInterface } from '../interfaces/domain-expert-interface';
import { SessionManager } from './session-manager';

export interface Request {
  id: string;
  type: string;
  sessionId: string;
  modality: string;
  content: any;
  metadata: Record<string, any>;
  timestamp: Date;
}

export interface Response {
  requestId: string;
  status: "success" | "error" | "pending";
  content: any;
  timestamp: Date;
  metadata?: Record<string, any>; // Added metadata field
}

export interface Task {
  id: string;
  type: string;
  content: any;
  context?: any;
  constraints?: {
    deadline?: Date;
    maxTokens?: number;
    priorityLevel?: "high" | "normal" | "low";
    maxRetries?: number;
  };
  metadata: Record<string, any>;
}

export interface TaskResult {
  taskId: string;
  success: boolean;
  result?: any;
  error?: string;
  metrics?: Record<string, any>;
}

export interface RequestHandler {
  handleRequest(request: Request): Promise<Response>;
  canHandle(request: Request): boolean;
}

export class Dispatcher {
  private readonly handlers: Map<string, RequestHandler[]> = new Map();
  private readonly inProgressRequests: Map<string, {
    request: Request,
    resolve: (value: Response) => void,
    reject: (reason: any) => void
  }> = new Map();
  private readonly contextManager: ContextManager;
  private readonly domainExpertInterface: DomainExpertInterface;
  private readonly sessionManager: SessionManager;

  constructor(
    contextManager: ContextManager,
    domainExpertInterface: DomainExpertInterface,
    sessionManager: SessionManager
  ) {
    this.contextManager = contextManager;
    this.domainExpertInterface = domainExpertInterface;
    this.sessionManager = sessionManager;
  }

  /**
   * Register a handler for a specific request type
   * @param requestType The type of request to handle
   * @param handler The handler for this request type
   */
  registerHandler(requestType: string, handler: RequestHandler): void {
    const existingHandlers = this.handlers.get(requestType) || [];
    this.handlers.set(requestType, [...existingHandlers, handler]);
    logger.info(`Registered handler for request type: ${requestType}`);
  }

  /**
   * Dispatch an incoming request to the appropriate handler
   * @param request The request to dispatch
   * @returns The response from the handler
   */
  async dispatch(request: Request): Promise<Response> {
    try {
      // If no ID is provided, generate one
      if (!request.id) {
        request.id = uuidv4();
      }

      // Set timestamp if not provided
      if (!request.timestamp) {
        request.timestamp = new Date();
      }

      logger.info(`Dispatching request ${request.id} of type ${request.type}`);

      // Validate session
      const sessionValid = await this.sessionManager.validateSession(request.sessionId);
      if (!sessionValid) {
        return {
          requestId: request.id,
          status: 'error',
          content: { error: 'Invalid or expired session' },
          timestamp: new Date()
        };
      }

      // Find handlers for this request type
      const handlers = this.handlers.get(request.type) || [];
      const eligibleHandlers = handlers.filter(handler => handler.canHandle(request));

      if (eligibleHandlers.length === 0) {
        logger.warn(`No handlers found for request type: ${request.type}`);
        return {
          requestId: request.id,
          status: 'error',
          content: { error: `No handler available for request type: ${request.type}` },
          timestamp: new Date()
        };
      }

      // Store in-progress request for cancellation support
      const responsePromise = new Promise<Response>((resolve, reject) => {
        this.inProgressRequests.set(request.id, { request, resolve, reject });
      });

      // Process request with the first eligible handler
      const handler = eligibleHandlers[0];

      try {
        // Get the session for context
        const session = await this.sessionManager.getSession(request.sessionId);

        if (!session) {
          throw new Error(`Invalid session: ${request.sessionId}`);
        }

        // Update context with this request
        await this.contextManager.updateContext(
          request.sessionId,
          session.activeWorkspaceId,
          {
            addMessages: [{
              id: request.id,
              role: 'user',
              content: JSON.stringify(request.content),
              timestamp: request.timestamp
            }]
          }
        );

        // Handle the request
        const response = await handler.handleRequest(request);

        // Update context with the response
        if (response.status === 'success') {
          await this.contextManager.updateContext(
            request.sessionId,
            session.activeWorkspaceId,
            {
              addMessages: [{
                id: `response-${request.id}`,
                role: 'assistant',
                content: JSON.stringify(response.content),
                timestamp: response.timestamp
              }]
            }
          );
        }

        // Resolve the promise with the response
        const requestInfo = this.inProgressRequests.get(request.id);
        if (requestInfo) {
          requestInfo.resolve(response);
          this.inProgressRequests.delete(request.id);
        }

        return response;
      } catch (error) {
        logger.error(`Error handling request ${request.id}`, error);

        const errorResponse: Response = {
          requestId: request.id,
          status: 'error',
          content: { error: (error as Error).message || 'Unknown error occurred' },
          timestamp: new Date()
        };

        // Resolve the promise with the error response
        const requestInfo = this.inProgressRequests.get(request.id);
        if (requestInfo) {
          requestInfo.resolve(errorResponse);
          this.inProgressRequests.delete(request.id);
        }

        return errorResponse;
      }
    } catch (error) {
      logger.error(`Failed to dispatch request`, error);
      return {
        requestId: request.id || 'unknown',
        status: 'error',
        content: { error: (error as Error).message || 'Unknown error occurred during dispatch' },
        timestamp: new Date()
      };
    }
  }

  /**
   * Route a task to a domain expert
   * @param expertType The type of domain expert to use
   * @param task The task to delegate
   * @returns The result from the domain expert
   */
  async delegateToExpert(expertType: string, task: Task): Promise<TaskResult> {
    try {
      logger.info(`Delegating task ${task.id || '(new)'} to expert: ${expertType}`);

      // If no ID is provided, generate one
      if (!task.id) {
        task.id = uuidv4();
      }

      // Check if the expert type is available
      const experts = await this.domainExpertInterface.listExperts();
      const expertExists = experts.some(expert => expert.type === expertType);

      if (!expertExists) {
        throw new Error(`Domain expert of type ${expertType} not available`);
      }

      // Delegate the task
      const taskId = await this.domainExpertInterface.delegateTask(expertType, task);

      // Start monitoring the task status
      const checkInterval = setInterval(async () => {
        try {
          const status = await this.domainExpertInterface.checkTaskStatus(taskId);

          if (status.state === 'completed' || status.state === 'failed' || status.state === 'cancelled') {
            clearInterval(checkInterval);
          }
        } catch (error) {
          logger.error(`Error checking task status for ${taskId}`, error);
          clearInterval(checkInterval);
        }
      }, 1000); // Check every second

      // Wait for task completion
      let currentStatus = await this.domainExpertInterface.checkTaskStatus(taskId);

      while (currentStatus.state !== 'completed' &&
        currentStatus.state !== 'failed' &&
        currentStatus.state !== 'cancelled') {
        // Wait before checking again
        await new Promise(resolve => setTimeout(resolve, 500));
        currentStatus = await this.domainExpertInterface.checkTaskStatus(taskId);
      }

      // Get the task result
      const result = await this.domainExpertInterface.getTaskResult(taskId);
      return result;
    } catch (error) {
      logger.error(`Failed to delegate task to expert ${expertType}`, error);
      return {
        taskId: task.id,
        success: false,
        error: (error as Error).message || 'Unknown error occurred during expert delegation'
      };
    }
  }

  /**
   * Cancel an in-progress request
   * @param requestId The ID of the request to cancel
   * @returns True if successfully cancelled
   */
  async cancelRequest(requestId: string): Promise<boolean> {
    try {
      logger.info(`Attempting to cancel request: ${requestId}`);

      const requestInfo = this.inProgressRequests.get(requestId);
      if (!requestInfo) {
        logger.warn(`Request ${requestId} not found or already completed`);
        return false;
      }

      // Reject the promise with a cancellation error
      requestInfo.reject(new Error('Request was cancelled'));

      // Remove from in-progress requests
      this.inProgressRequests.delete(requestId);

      return true;
    } catch (error) {
      logger.error(`Failed to cancel request ${requestId}`, error);
      return false;
    }
  }
}
